File: PyEngineering/06_atomic_file_store/06_atomic_file_store_explanation.py
"""
06 — Atomic File Store
=======================

WHAT
----
Build `AtomicFileStore`, a small on-disk key/value store where every write is
*atomic from the reader's point of view*: a concurrent reader (or a crash)
must never observe a partially-written file. Callers get either the old
content or the fully-written new content — never a truncated or torn mix of
both.

WHY THIS MATTERS
----------------
`open(path, "w")` truncates the file *before* your first byte is written.
Anyone who opens that path between your `open()` and your final `close()`
sees a zero-length or half-written file. A crash (power loss, OOM-killer,
`kill -9`, the containerd sidecar getting rescheduled mid-deploy) at that
moment leaves the file corrupted *on disk*, not just in a buffer — there is
no "undo" on restart.

This is not a hypothetical: config files, checkpoint files, on-disk caches,
lock/state files for daemons, and single-file "poor man's databases" all rely
on this pattern in real systems. `sqlite3` itself uses a close cousin of it
(WAL / rollback journal) for the same reason problem 09/10 need transactions
at all: torn writes are unacceptable.

THE CORE TRICK
---------------
1. Write the *new* content to a temporary file **in the same directory** as
   the target (same filesystem/mount — required, see PITFALLS).
2. `flush()` + `os.fsync(tmp_fd)` the temp file's contents to durable
   storage — without this, `os.replace` durably renames a file whose
   *contents* may still only exist in the OS page cache.
3. `os.replace(tmp_path, target_path)` — on POSIX this is `rename(2)`, which
   is guaranteed atomic: any concurrent `open()` on `target_path` either
   sees the fully-old file or the fully-new file, never a mix, because a
   directory-entry swap is a single filesystem operation. `os.replace` is
   also atomic on Windows (unlike `os.rename`, which raises if the
   destination exists there) — that's precisely why the stdlib docs
   recommend `os.replace` over `shutil.move` (which may fall back to a
   non-atomic copy+delete across filesystems) or `os.rename` (platform-
   dependent overwrite semantics) for this use case.
4. (Optional but recommended for "survive a real crash, not just a
   process crash") `os.fsync` the *directory* file descriptor after the
   rename, so the rename itself is durable and not just the file's data.

WHAT ATOMIC RENAME DOES **NOT** GIVE YOU
-----------------------------------------
- It does not make multi-file updates atomic (only a single rename is
  atomic; if you need N files to change together, that's a mini-transaction
  problem — write a manifest file atomically instead, or see problem 10).
- It does not survive the temp file and target being on different
  filesystems/mounts (`os.replace` across devices raises `OSError` — this is
  why the temp file MUST be created in the same directory as the target).
- It does not remove the *durability* gap between "data written" and "fsync
  called" — you must fsync the temp file's data before the rename, or a
  crash between `close()` and the rename can still leave the *new* file
  (post-crash) with zero bytes because the rename itself is fine but the
  data it points to was never flushed from cache to disk.

SPEC
----
Implement `AtomicFileStore`:

    class AtomicFileStore:
        def __init__(self, root: Path, *, file_mode: int = 0o644) -> None: ...

        def write_bytes(self, key: str, data: bytes) -> None:
            "Atomically create-or-replace `root/key` with `data`."

        def write_text(self, key: str, text: str, *, encoding: str = "utf-8") -> None:
            "Same as write_bytes but for str content."

        def read_bytes(self, key: str) -> bytes: ...
        def read_text(self, key: str, *, encoding: str = "utf-8") -> str: ...

        def delete(self, key: str) -> None:
            "Remove the file for `key`. Missing key -> FileNotFoundError."

        def exists(self, key: str) -> bool: ...

Rules the implementation must satisfy:

- `key` must not escape `root` (no `..`, no absolute paths, no symlink
  traversal out of `root`) — validate and raise `ValueError` on a bad key.
  This is the same class of bug as a path-traversal vulnerability in a web
  upload handler; a "local file store" often becomes exactly that handler.
- The temp file used for staging must live in the *same directory* as the
  target file (use `tempfile.mkstemp(dir=root, ...)` or equivalent) — never
  `tempfile.gettempdir()` — otherwise the final `os.replace` can silently
  become non-atomic (cross-device) or outright fail.
- On any exception during the write (encode error, disk full, etc.) the temp
  file must be cleaned up — the target file must be left untouched.
- `file_mode` should be applied to the final file (accounting for the
  process `umask`, which `mkstemp` applies restrictively by default — note
  in your solution's comments what `mkstemp`'s default mode is and why you
  must `os.chmod` explicitly if you want anything looser).
- Must work correctly if called from multiple threads/processes targeting
  *different* keys concurrently (no shared mutable state beyond the
  filesystem itself).

ACCEPTANCE CRITERIA
--------------------
- Writing then reading a key round-trips exactly.
- Overwriting an existing key never leaves a truncated/partial file visible
  — this is what the test suite exercises by simulating a failure mid-write
  and asserting the *original* content is still intact afterward.
- No leftover `tmp*` files remain in `root` after a successful OR a failed
  write.
- Path-traversal keys (`"../escape"`, `"/etc/passwd"`, `"a/../../b"`) are
  rejected with `ValueError` before any filesystem write is attempted.
- Correct file permissions on the resulting file.
"""

from __future__ import annotations

from pathlib import Path


class AtomicFileStore:
    """A directory-backed key/value store with crash-safe atomic writes."""

    def __init__(self, root: Path, *, file_mode: int = 0o644) -> None:
        # TODO: store root (resolved to an absolute path), create it if
        # missing, store file_mode.
        raise NotImplementedError

    def _resolve(self, key: str) -> Path:
        """Validate `key` and return the absolute path it maps to.

        TODO: reject empty keys, absolute-path keys, and any key whose
        resolved path escapes `self.root` (use Path.resolve() + a
        relative_to() check, or is_relative_to() on 3.9+/3.12).
        """
        raise NotImplementedError

    def write_bytes(self, key: str, data: bytes) -> None:
        """Atomically create-or-replace the file for `key` with `data`.

        TODO:
        - resolve/validate key
        - mkstemp() in the same directory
        - write + flush + fsync the temp file
        - chmod to self.file_mode
        - os.replace(tmp, target)
        - fsync the containing directory
        - clean up the temp file on any exception path
        """
        raise NotImplementedError

    def write_text(self, key: str, text: str, *, encoding: str = "utf-8") -> None:
        """TODO: encode and delegate to write_bytes."""
        raise NotImplementedError

    def read_bytes(self, key: str) -> bytes:
        """TODO: resolve/validate key, read and return file contents."""
        raise NotImplementedError

    def read_text(self, key: str, *, encoding: str = "utf-8") -> str:
        """TODO: decode read_bytes()."""
        raise NotImplementedError

    def delete(self, key: str) -> None:
        """TODO: resolve/validate key, unlink. Missing file -> let
        FileNotFoundError propagate (do not swallow it)."""
        raise NotImplementedError

    def exists(self, key: str) -> bool:
        """TODO: resolve/validate key, return whether the file exists."""
        raise NotImplementedError


# ---------------------------------------------------------------------------
# Hints
# ---------------------------------------------------------------------------
# - `tempfile.mkstemp(dir=..., prefix=..., suffix=...)` returns (fd, path);
#   wrap the fd with `os.fdopen(fd, "wb")` so you get a normal file object
#   you can `.write()` to, or use `os.write(fd, data)` + `os.close(fd)`
#   directly — either is fine, just be consistent about who closes what.
# - `Path.is_relative_to()` (3.9+) is the clean way to check containment
#   after `Path.resolve()`.
# - `contextlib.suppress(FileNotFoundError)` is handy when cleaning up a
#   temp file that might already be gone.
# - Directory fsync: `fd = os.open(dir_path, os.O_RDONLY); os.fsync(fd);
#   os.close(fd)` — this is the step people forget; without it the rename
#   itself might not survive a power loss even though the file content did.
#
# Pitfalls
# --------
# - Creating the temp file with `tempfile.NamedTemporaryFile()` (default
#   directory) instead of `dir=root` — breaks atomicity if `/tmp` is a
#   different filesystem/mount than `root` (extremely common on Linux where
#   `/tmp` is tmpfs).
# - Forgetting `os.fsync` on the temp file before `os.replace` — the rename
#   is atomic but the *data* it points to might not be durable yet.
# - Using `shutil.move` — falls back to a non-atomic copy across devices,
#   silently defeating the whole point of this exercise.
# - Not validating keys — a naive `root / key` join lets `key="../../etc/cron.d/x"`
#   write outside the store entirely.
#
# Stretch goals
# -------------
# - Add `write_bytes_if_unchanged(key, data, expected_old: bytes | None)` —
#   an optimistic-concurrency variant that raises if the current on-disk
#   content doesn't match `expected_old` (read-check-write race is still
#   possible without a lock; document that honestly rather than pretending
#   it's solved — true compare-and-swap needs problem 10's techniques or a
#   file lock).
# - Add a `list_keys() -> Iterator[str]` that walks `root`.
# - Make the store survive being handed a `root` that doesn't exist yet by
#   creating parent directories, and reason in a comment about whether that
#   directory-creation step itself needs a directory fsync.

File: GoEngineering/06_atomic_file_store/explanation/06_atomic_file_store_explanation.go
/*
Package filestore — Atomic File Store (Problem 06)

WHAT WE'RE BUILDING

A small library that writes files to disk *atomically*: readers never observe a
partially-written file, and a crash (power loss, SIGKILL, panic mid-write) never
leaves a corrupt file where a good one used to be. This is the pattern behind
every "safe config write", local database WAL checkpoint, and CLI tool that
edits a file in place (kubectl config, git's index, Terraform state, etc).

# WHY IT MATTERS IN REAL SYSTEMS

Naively doing:

	f, _ := os.Create(path)
	f.Write(data)
	f.Close()

has two failure windows:
 1. Between truncate-on-create and the write completing, any concurrent
    reader (or a crash) sees a zero-length or half-written file.
 2. Even after Close returns, the data may still be sitting in the OS page
    cache, not on durable storage — a power cut can lose it entirely, or
    leave the file's metadata (size) updated but content stale/torn on some
    filesystems.

The standard fix, used by etcd, Consul, containerd, and most Unix editors
(vim, sed -i) is the "temp file + fsync + rename" pattern:

 1. Create a temp file in the SAME DIRECTORY as the destination (same
    filesystem/mount — rename() is only atomic within one filesystem).
 2. Write the full contents to the temp file.
 3. fsync the temp file's data to disk (forces the write out of the page
    cache onto durable storage).
 4. Close the temp file.
 5. rename(tempPath, destPath) — POSIX guarantees this is atomic: any
    concurrent open() of destPath sees either the fully-old or fully-new
    file, never a mix, never a gap.
 6. fsync the CONTAINING DIRECTORY. On most POSIX filesystems the rename
    itself is a metadata change to the directory, which can also be lost on
    crash unless the directory's fd is fsynced. Without this step you can
    legitimately end up, after a crash, with the new file's data safely on
    disk but the directory entry still pointing at the old inode (or a
    torn/missing entry) — the "rename didn't survive the crash" bug class.

CONCEPTS COVERED
  - os.CreateTemp for collision-free temp file naming
  - File.Sync (fsync) and why Close() alone is not durable
  - os.Rename atomicity guarantees and its same-filesystem requirement
  - Directory fsync (opening a directory with os.Open and calling Sync)
  - File permission handling (temp files default to 0600; we need to set the
    final mode explicitly since CreateTemp doesn't respect a requested mode)
  - Cleanup-on-error (must remove the temp file if anything fails before
    rename, or you leak files on every error path)
  - context.Context cancellation mid-write
  - Concurrency safety: multiple goroutines calling Write concurrently for
    the same store must not corrupt each other's temp files or race on the
    final rename

REQUIREMENTS / SPEC

Implement a `Store` type bound to a directory, with:

	NewStore(dir string) (*Store, error)
	    Validates dir exists and is a directory. Returns *Store.

	(*Store) WriteFile(ctx context.Context, name string, data []byte, perm os.FileMode) error
	    Atomically writes `data` to <dir>/<name> using the temp+fsync+rename
	    pattern described above. Must:
	      - Create the temp file in the same directory as the target (NOT
	        os.TempDir()).
	      - Respect ctx cancellation: check ctx.Err() before starting and
	        before the (relatively slow) fsync/rename steps; if canceled,
	        clean up the temp file and return ctx.Err().
	      - Set the final file's permission bits to `perm` (temp files are
	        created 0600 by os.CreateTemp; you must Chmod before or after
	        rename to get the caller's requested mode onto the final path).
	      - fsync the temp file before renaming.
	      - fsync the containing directory after renaming.
	      - On ANY error prior to a successful rename, remove the leftover
	        temp file (best-effort — log/ignore removal errors, don't mask the
	        original error).
	      - Be safe for concurrent callers writing DIFFERENT names in the same
	        store (they must not interfere with each other).

	(*Store) ReadFile(name string) ([]byte, error)
	    Plain read of <dir>/<name>. Wrap os.ErrNotExist-producing errors so
	    callers can `errors.Is(err, os.ErrNotExist)`.

ACCEPTANCE CRITERIA
  - WriteFile never leaves a *.tmp* file behind on success.
  - A concurrent reader opening the destination path during a WriteFile call
    always sees either the fully old content or the fully new content — this
    is testable by racing many goroutines writing increasing-size payloads
    against a goroutine repeatedly opening+reading the file, and asserting
    the length always matches a length that was actually written.
  - Canceling ctx before WriteFile completes returns context.Canceled/
    DeadlineExceeded and leaves the destination file untouched (old content
    or nonexistent), with no leaked temp file.
  - Final file has the requested permission bits (masked by umask, as usual
    on POSIX).
  - `go test -race` is clean.
*/
package filestore

import (
	"context"
	"os"
)

// Store writes files atomically into a single fixed directory.
type Store struct {
	// TODO: store the target directory (absolute path is safest — resolve
	// with filepath.Abs in NewStore so later os.Open(s.dir) calls are
	// unambiguous regardless of the process's current working directory).
}

// NewStore returns a Store rooted at dir. dir must already exist and be a
// directory; NewStore does not create it.
func NewStore(dir string) (*Store, error) {
	// TODO:
	//  1. os.Stat(dir); if error, wrap and return.
	//  2. Confirm it's a directory (info.IsDir()); if not, return a clear
	//     error.
	//  3. Resolve to an absolute path (filepath.Abs) and store it.
	panic("TODO: implement NewStore")
}

// WriteFile atomically replaces (or creates) <dir>/<name> with data, using
// the temp-file + fsync + rename pattern. perm sets the final file's mode.
//
// See the package doc comment above for the full durability contract this
// must satisfy.
func (s *Store) WriteFile(ctx context.Context, name string, data []byte, perm os.FileMode) error {
	// TODO:
	//  1. Check ctx.Err() up front — bail early if already canceled.
	//  2. os.CreateTemp(s.dir, ".<name>.tmp-*") to get a sibling temp file
	//     with a random suffix (collision-free for concurrent writers).
	//  3. Write data to it in full (Write can short-write; loop or use
	//     io.Copy/Write with a check, though for a single []byte a direct
	//     tmp.Write(data) length-checked against len(data) is enough).
	//  4. tmp.Sync() — fsync the file data.
	//  5. tmp.Chmod(perm) (or os.Chmod after close — either works, but doing
	//     it before Close means one fewer syscall pair) then tmp.Close().
	//  6. Check ctx.Err() again before the rename (cheap insurance against
	//     racing a long write against a very tight deadline).
	//  7. os.Rename(tmpPath, destPath).
	//  8. Open the containing directory and Sync() it, then Close() it.
	//  9. On any error in steps 2-7, os.Remove(tmpPath) best-effort before
	//     returning the original error.
	panic("TODO: implement WriteFile")
}

// ReadFile reads <dir>/<name> in full. Errors that indicate the file does
// not exist must satisfy errors.Is(err, os.ErrNotExist).
func (s *Store) ReadFile(name string) ([]byte, error) {
	// TODO: os.ReadFile(filepath.Join(s.dir, name)), wrapping errors with
	// fmt.Errorf("%w: ...") so os.ErrNotExist is preserved through errors.Is.
	panic("TODO: implement ReadFile")
}

/*
HINTS
  - filepath.Join(s.dir, name) for both the temp file's directory argument
    and the final destination — never write into os.TempDir(), it's very
    likely a different filesystem/mount, which makes os.Rename fail with
    "invalid cross-device link" (EXDEV) instead of being atomic.
  - os.CreateTemp's pattern argument supports a "*" wildcard that gets
    replaced with a random string; put it in the middle of the name (e.g.
    ".config.json.tmp-*") so the temp file sorts near the real file and is
    easy to spot/clean up if something goes wrong.
  - Don't forget: os.CreateTemp files are mode 0600 regardless of the
    destination's desired permissions — you must Chmod explicitly.
  - The directory fsync step is the one everyone forgets. It matters: without
    it, a crash right after rename() can lose the directory-entry update on
    some filesystems/configurations even though the file's own data is safe.

COMMON PITFALLS
  - Writing the temp file to os.TempDir() (different filesystem -> rename
    fails or, worse, silently falls back to non-atomic copy+delete on some
    libraries).
  - Forgetting to clean up the temp file on error paths — leaks accumulate
    silently in the target directory over time.
  - Calling Close() and assuming durability — Close() does not imply fsync
    on most OSes; you must call Sync() explicitly before renaming.
  - Chmod-ing after rename on a path that no longer matches perm because a
    concurrent writer already replaced it with a *different* new version —
    prefer chmod-then-rename (chmod the temp file, so the rename is what
    publishes the final permission atomically with the content).
  - Ignoring ctx cancellation entirely, or checking it only once at the very
    top (a write of a large payload can take a while; check again before the
    slow fsync/rename phase too).

STRETCH GOALS
  - Add a `WriteReader(ctx, name string, r io.Reader, perm) error` variant
    that streams into the temp file via io.Copy instead of taking a []byte,
    for large payloads that shouldn't be buffered fully in memory.
  - Add an optional checksum: write a sidecar `<name>.sha256` atomically
    alongside the content (two atomic writes — think about what "atomic" now
    means for the *pair*, and document the limitation).
  - Support Linux's renameat2(RENAME_NOREPLACE) via golang.org/x/sys/unix to
    offer a CreateFileExclusive that fails if the destination already
    exists, instead of always overwriting.
  - Benchmark WriteFile with and without the directory fsync to see the real
    cost of durability (b.ReportMetric with ns/op) — discuss when you'd skip
    it (e.g. bulk-loading data you can regenerate) vs never skip it (financial
    ledgers, embedded databases).
*/
